index.js

```diff
@@ -83,6 +94,116 @@ function attachToBody (options, req, reply, next) {
+  return tokens.some(token => ['[', ']', '{}'].includes(token)) ? tokens : []
+}
+
+function validateSerializedKey (tokens, value) {
+  let isValid = true
+
+  // first token can't be '[', ']', or '{}'
+  if (['[', ']', '{}'].includes(tokens[0])) isValid = false
+  // only a single {} special ending is allowed
+  else if (tokens.filter(t => t === '{}').length > 1) isValid = false
+  // {}, if present, must be the last token
+  else if (tokens.find(t => t === '{}') && tokens[tokens.length - 1] !== '{}') isValid = false
+
+  let hasOpenBracket = false
+  for (let i = 1; i < tokens.length; i++) {
+    const token = tokens[i]
+    const nextToken = tokens[i + 1]
+    // invalid closing bracket
+    if (!hasOpenBracket && token === ']') {
+      isValid = false
+      break
+    }
+    // invalid opening bracket
+    if (hasOpenBracket && token === '[') {
+      isValid = false
+      break
+    }
+    hasOpenBracket = token === '[' || (hasOpenBracket && token !== ']')
+    // orphan opening bracket
+    if (hasOpenBracket && nextToken === undefined) isValid = false
+    // {} is not allowed within brackets
+    if (hasOpenBracket && token === '{}') isValid = false
+    // after a closing bracket, only an opening bracket or {} are allowed
+    if (token === ']' && nextToken && !['[', '{}'].includes(nextToken)) {
+      isValid = false
+      break
+    }
+  }
+
+  // when the {} special ending is supplied, the value must be valid JSON
+  if (isValid && tokens[tokens.length - 1] === '{}') {
+    try {
+      secureJSON.parse(value)
+    } catch (e) {
+      return 'the {} special ending was used, but value is not a valid JSON'
+    }
+  }
+
+  if (tokens.includes('__proto__')) {
+    return '__proto__ is not allowed as a key name'
+  }
+
+  if (tokens.includes('constructor')) {
+    return 'constructor is not allowed as a key name'
+  }
+
+  return isValid ? null : 'invalid serialized key'
+}
+
+function getSerializedKeySegments (options, body, tokens, value) {
+  const segments = [{ parentRef: body, key: tokens[0] }]
+  const parseArrays = options.enableSpecialNotationKeys !== 'objectsOnly'
+
+  for (let i = 1; i < tokens.length; i++) {
+    const lastSegment = segments[segments.length - 1]
+    const token = tokens[i]
+    const prevToken = tokens[i - 1]
+    let parentRef
+    let key
+
+    if (token === ']' && prevToken === '[') {
+      parentRef = lastSegment.parentRef[lastSegment.key] ?? (parseArrays ? [] : {})
+      key = parseArrays ? parentRef.length : Object.keys(parentRef).length
+    } else if (parseArrays && token === ']' && isValidIndex(prevToken)) {
+      parentRef = lastSegment.parentRef[lastSegment.key] ?? (parseArrays ? [] : {})
+      key = +prevToken
+    } else if (token === ']') {
+      parentRef = lastSegment.parentRef[lastSegment.key] ?? {}
+      key = prevToken
+    } else if (token === '{}') {
+      value = secureJSON.parse(value)
+    }
+
+    if (parentRef) segments.push({ parentRef, key })
```

Comment:
```suggestion
    parentRef && segments.push({ parentRef, key })
```

Reply:
I've nothing against using short circuit to replace inline ifs, but in this case, I don't see the real benefit. I even [benchmarked it](https://perf.link/#eyJpZCI6IndjcXoyZDM5OTFwIiwidGl0bGUiOiJJZiB2cyBzaG9ydC1jaXJjdWl0IiwiYmVmb3JlIjoibGV0IGZvbyA9IHRydWU7IiwidGVzdHMiOlt7Im5hbWUiOiJJZiIsImNvZGUiOiJpZiAoZm9vKSBjb25zb2xlLmxvZygnYmFyIScpOyIsInJ1bnMiOls3MDAwMCw0ODAwMCw5NjAwMCw3MTAwMCw3NDAwMCw3MTAwMCwxMDIwMDAsMTQ1MDAwLDUwMDAsMTc0MDAwLDExMDAwMCwxMjEwMDAsODgwMDAsMjUwMDAsMTI3MDAwLDE0NTAwMCwxMjMwMDAsMTczMDAwLDgwMDAwLDE4NDAwMCwxNDkwMDAsNjgwMDAsNTEwMDAsMTYwMDAsMTIxMDAwLDEwMzAwMCwxMjEwMDAsNjEwMDAsMTYzMDAwLDQ2MDAwLDE4NTAwMCw1OTAwMCwyMjAwMCw1MDAwLDEzNjAwMCw3NDAwMCw4MjAwMCwxMDAwLDE2MDAwLDc4MDAwLDM1MDAwLDk2MDAwLDE1MDAwLDQ5MDAwLDEwMDAsMTAwMCw2MTAwMCwyMTcwMDAsMTE0MDAwLDUyMDAwLDE1MDAwMCw2NTAwMCw2ODAwMCwyMDAwLDEwMDAsNTMwMDAsMjAwMCwxMDAwMDAsMTUxMDAwLDQwMDAwLDQzMDAwLDUzMDAwLDEzOTAwMCwxMDUwMDAsOTQwMDAsNzAwMDAsNTkwMDAsMTEwMDAwLDkxMDAwLDQ1MDAwLDMwMDAsMTg2MDAwLDY3MDAwLDEwMDAsNzAwMDAsMTMzMDAwLDMxMDAwLDU3MDAwLDY0MDAwLDg4MDAwLDQwMDAwLDQ2MDAwLDEyNzAwMCwxMjEwMDAsNzcwMDAsOTcwMDAsNzAwMCwzNzAwMCw4MDAwMCwyNDAwMCwxNDEwMDAsNjcwMDAsMjEwMDAwLDEwMDAsNzEwMDAsNzkwMDAsMTIwMDAwLDc1MDAwLDIxOTAwMCwxNzAwMDBdLCJvcHMiOjgxODAwfSx7Im5hbWUiOiJTaG9ydC1jaXJjdWl0IiwiY29kZSI6ImZvbyAmJiBjb25zb2xlLmxvZygnYmFyIScpOyIsInJ1bnMiOlsxODAwMCwxNjQwMDAsMzAwMDAsMzAwMCw3NzAwMCw3MTAwMCwyMjAwMDAsNzgwMDAsMTQ3MDAwLDEwMDAsNTAwMDAsMzUwMDAsODUwMDAsMjUwMDAsMTY3MDAwLDEzODAwMCw2NzAwMCw0NDAwMCw2NjAwMCw2MDAwLDE0MDAwMCwxOTAwMDAsMTgwMDAsNzEwMDAsMTIzMDAwLDQ1MDAwLDExNjAwMCwxNzgwMDAsOTIwMDAsMTY5MDAwLDEwODAwMCwyMzAwMCwzMTAwMCw3NjAwMCw5MDAwLDEwMDAsMTg1MDAwLDEwMDAsNTcwMDAsMTI2MDAwLDcwMDAwLDEwMTAwMCwxMDQwMDAsMTAwMCwxNTgwMDAsOTgwMDAsNTUwMDAsMzEwMDAsNTcwMDAsMzcwMDAsNzEwMDAsMTQ2MDAwLDQwMDAwLDI3MDAwLDU0MDAwLDE1OTAwMCwyNzAwMCwxOTIwMDAsMzMwMDAsNjcwMDAsMTU3MDAwLDEwMDAwLDMwMDAwLDEwMDAsODcwMDAsMTUwMDAsMjYwMDAsNDAwMDAsMTM0MDAwLDEwMDAsMTAzMDAwLDMzMDAwLDUwMDAsMTA3MDAwLDEzMzAwMCwyNTAwMCwxMzcwMDAsMTIwMDAsMTg1MDAwLDE1MjAwMCwyMTAwMDAsMjMwMDAsODYwMDAsMjAwMDAsNTkwMDAsNzgwMDAsMTIwMDAwLDY4MDAwLDEwOTAwMCw0NzAwMCw5NjAwMCwxODQwMDAsMTAwMCwxMDEwMDAsNzkwMDAsNjUwMDAsODgwMDAsNDAwMCwyMTEwMDAsMTIxMDAwXSwib3BzIjo3OTQyMH1dLCJ1cGRhdGVkIjoiMjAyMy0wNS0xMVQyMDoxODoxMC44MzBaIn0%3D), but the results are not conclusive. If I rerun the benchmark multiple times, sometimes `if` wins by a slim margin, other times short circuit is slightly faster. So, I don't see why we should change it.

Reply:
Then please add some curly brackets please ;)